Added Settings to deal with newer settings & add callback threads

Daeda88 opened the pull request:
Final one for now:

This one updates Firestore/Database to provide settings. This includes fixes for deprecated Persistency settings (though not on Java as that is not tracked yet) and adds support on iOS/Android to set custom callback dispatchers. Otherwise, Firebase internally still goes to the main thread even when running in a background thread.

Daeda88 commented:
@nbransby this is now tracking master, which means the JVM split for firestore is tracked. This PR can now be properly reviewed

Daeda88 commented:
Updated tracking again @nbransby

nbransby commented on firebase-database/src/commonMain/kotlin/dev/gitlive/firebase/database/database.kt:
```diff
@@ -32,13 +32,25 @@ expect fun Firebase.database(app: FirebaseApp): FirebaseDatabase
     fun reference(path: String): DatabaseReference
     fun reference(): DatabaseReference
-    fun setPersistenceEnabled(enabled: Boolean)
+    fun setSettings(settings: Settings)
```
why not just    
`fun setSettings(persistenceEnabled: Boolean = false, persistenceCacheSizeBytes: Long? = null)`
What the point of making everyone write the boilerplate
`database.setSettings(FirebaseDatabase.Settings.createSettings(persistenceEnabled = true, persistenceCacheSizeBytes = 2048))`
?

nbransby commented:
Looking at the official android sdk, the methods are defined as follows:
https://firebase.google.com/docs/reference/android/com/google/firebase/database/FirebaseDatabase#setPersistenceCacheSizeBytes(long)
Any reason not to match this?

nbransby commented:
This PR inspired to write a section on compatibility with the official firebase android sdk here: https://github.com/GitLiveApp/firebase-kotlin-sdk?tab=readme-ov-file#compatibility-with-the-official-firebase-android-sdk

Can we apply those points to this PR?

Daeda88 commented:
Yep, will track/correct when I have the time

Daeda88 commented:
Changes now track the Android API. I did leave the LocalCacheSettings as a sealed class as opposed to an interface, so it can both be constructed through a direct constructor as well as through builders as they exist on Android.

nbransby commented on firebase-database/src/iosMain/kotlin/dev/gitlive/firebase/database/database.kt:
```diff
@@ -64,6 +67,14 @@ actual class FirebaseDatabase internal constructor(val ios: FIRDatabase) {
+        ios.setPersistenceCacheSizeBytes(cacheSizeInBytes.toULong())
+    }
+
+    fun setCallbackQueue(callbackQueue: dispatch_queue_t) {
+        ios.callbackQueue = callbackQueue
+    }
```
lets not start adding coverage for platform specific APIs, let users use .ios

nbransby commented on firebase-firestore/src/commonTest/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -100,7 +100,11 @@ class FirebaseFirestoreTest {
         firestore = Firebase.firestore(app).apply {
             useEmulator(emulatorHost, 8080)
-            setSettings(persistenceEnabled = false)
+            settings = firestoreSettings(settings) {
+                cacheSettings = memoryCacheSettings {
+                    gcSettings = memoryEagerGcSettings { }
+                }
+            }
```
this syntax is bizarre, why not use the same pattern you used for encodeSettings? namely  `setSettings { ... }`  and put all the setting functions in the builderscope instead of at the package level

nbransby replied:
Oh shit 🤣

nbransby replied:
I'm starting to get the impression they have less kotlin experience on the android sdk team that us 😅

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+fun memoryCacheSettings(builder: LocalCacheSettings.Memory.Builder.() -> Unit): LocalCacheSettings.Memory =
+    LocalCacheSettings.Memory.newBuilder().apply(builder).build()
+
+fun memoryEagerGcSettings(builder: GarbageCollectorSettings.Eager.Builder.() -> Unit) =
+    GarbageCollectorSettings.Eager.newBuilder().apply(builder).build()
+
+fun memoryLruGcSettings(builder: GarbageCollectorSettings.LRUGC.Builder.() -> Unit) =
+    GarbageCollectorSettings.LRUGC.newBuilder().apply(builder).build()
+
+fun persistentCacheSettings(builder: LocalCacheSettings.Persistent.Builder.() -> Unit) =
+    LocalCacheSettings.Persistent.newBuilder().apply(builder).build()
```
strange having these at the package level, best moved into a scope, see https://github.com/GitLiveApp/firebase-kotlin-sdk/pull/450/files#r1568741186

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -27,11 +30,56 @@ expect class FirebaseFirestore {
+val FirestoreSettings.Companion.CACHE_SIZE_UNLIMITED get() = -1L
+val FirestoreSettings.Companion.DEFAULT_HOST get() = "firestore.googleapis.com"
+
+expect class FirestoreSettings {
```
rename FirebaseFirestoreSettings to match android

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -27,11 +30,56 @@ expect class FirebaseFirestore {
     suspend fun enableNetwork()
 }
 
+val FirestoreSettings.Companion.CACHE_SIZE_UNLIMITED get() = -1L
```
move inside companion object for binary compat with android

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -27,11 +30,56 @@ expect class FirebaseFirestore {
 }
 
+val FirestoreSettings.Companion.CACHE_SIZE_UNLIMITED get() = -1L
+val FirestoreSettings.Companion.DEFAULT_HOST get() = "firestore.googleapis.com"
```
I dont see this property on android?

Daeda88 replied:
https://github.com/firebase/firebase-android-sdk/blob/1d717ea6c7f1feb3219b254008212e8e73b9ce18/firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestoreSettings.java#L33

nbransby replied:
https://stackoverflow.com/questions/17035271/what-does-hide-mean-in-the-android-source-code

Daeda88 replied:
Ill make it internal only

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -27,11 +30,56 @@ expect class FirebaseFirestore {
+
+    companion object {}
+
+    interface Builder {
```
make this a class with the constructors available here to match android:
https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/FirebaseFirestoreSettings.Builder

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+package dev.gitlive.firebase.firestore
+
+sealed class LocalCacheSettings {
```
make sealed interface to android

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+
+sealed class LocalCacheSettings {
+
+    internal companion object {
+        // Firestore cache defaults to 100MB
+        const val DEFAULT_CACHE_SIZE = 100L*1024L*1024L
+    }
```
dont see this in the android sdk?

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+}
+
```
can we add `typealias MemoryCacheSettings = LocalCacheSettings.Memory`  and for the others for android sdk binary compat?

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+            override fun build(): Persistent = Persistent(sizeBytes)
+        }
+    }
+    data class Memory(val garbaseCollectorSettings: GarbageCollectorSettings) : LocalCacheSettings() {
```
make constructor internal or private

nbransby commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/LocalCacheSettings.kt:
```diff
@@ -0,0 +1,90 @@
+
+        interface Builder {
+
+            var gcSettings: GarbageCollectorSettings
```
rename garbageCollectorSettings for android sdk binary compat

Daeda88 replied:
Its called `gcSettings` though: https://firebase.google.com/docs/reference/kotlin/com/google/firebase/firestore/MemoryCacheSettings.Builder

nbransby replied:
sorry yes got confused with the name MemoryCacheSettings where it has a different name 🤯

nbransby reviewed (changes requested):
I didn't comment on every change needed to comply with https://github.com/GitLiveApp/firebase-kotlin-sdk?tab=readme-ov-file#compatibility-with-the-official-firebase-android-sdk but there should be enough comments there so you can apply the same changes to the remaining classes, fields and functions

nbransby commented:
Also I used the term binary compat a lot when technically I mean source compatibility

Daeda88 commented on firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:
```diff
@@ -27,11 +30,56 @@ expect class FirebaseFirestore {
+
+expect fun firestoreSettings(settings: FirestoreSettings? = null, builder: FirestoreSettings.Builder.() -> Unit): FirestoreSettings
+
+@Deprecated("Use dev.gitlive.firebase.firestore instead", replaceWith = ReplaceWith("setSettings(FirebaseFirestore.Settings.create())"))
+fun FirebaseFirestore.setSettings(
+    persistenceEnabled: Boolean? = null,
+    sslEnabled: Boolean? = null,
+    host: String? = null,
+    cacheSizeBytes: Long? = null,
+) {
+    settings = firestoreSettings {
+        this.sslEnabled = sslEnabled ?: true
+        this.host = host ?: FirestoreSettings.DEFAULT_HOST
+        this.cacheSettings = if (persistenceEnabled != false) {
+            LocalCacheSettings.Persistent(cacheSizeBytes ?: FirestoreSettings.CACHE_SIZE_UNLIMITED)
+        } else {
+            val cacheSize = cacheSizeBytes ?: FirestoreSettings.CACHE_SIZE_UNLIMITED
+            val garbageCollectionSettings = if (cacheSize == FirestoreSettings.CACHE_SIZE_UNLIMITED) {
+                GarbageCollectorSettings.Eager
+            } else {
+                GarbageCollectorSettings.LRUGC(cacheSize)
+            }
+            LocalCacheSettings.Memory(garbageCollectionSettings)
+        }
+    }
+}
+
 @PublishedApi
```
FirebaseFirestore is expect/actuals so it cant have default implementation. Omving to the same NativeWrapper pattern used elsewhere now

Daeda88 commented:
Updated with fixes for remarks. Also moved the `IllegalStateException` thrown when attempting to set new settings after using Firestore into the common code, so that it may be caught properly (before it was impossible to do so on iOS)

nbransby commented:
> Also moved the `IllegalStateException` thrown when attempting to set new settings after using Firestore into the common code, so that it may be caught properly (before it was impossible to do so on iOS)

Let's not start copying business logic such as this from the native SDKs as its causes a maintenance headache -  you can see [here](https://github.com/firebase/firebase-ios-sdk/blob/d58541efcaee7816b4240a525a20f101a167bdf4/Firestore/core/src/api/firestore.cc#L122) the logic is more complex. Lets leverage our dependency on the native SDKs and let them do the validation.

The fact you can't catch it on iOS is a good thing, an assertion like this represents a [programming error](https://github.com/firebase/firebase-ios-sdk/blob/d58541efcaee7816b4240a525a20f101a167bdf4/Firestore/core/src/util/exception.h#L20), not expected runtime behavior.

Daeda88 commented:
Removed then

nbransby commented:
look like the build failed:
`e: file:///home/runner/work/firebase-kotlin-sdk/firebase-kotlin-sdk/firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:47:17 Unresolved reference: isConfigured
`

Daeda88 commented:
Whoops, dont do things quickly in the morning :')

nbransby commented:
Js test still failing 😅

Daeda88 commented:
Source changes pointed to an actual bug. Fixed now
